### PR TITLE
[ROCm][K3] Extend FP8 asm MLA prefill to non-divisor small head counts

### PR DESCRIPTION
**Pull request by xiaohuguo2023**
## Context

The AITER FP8 MLA **prefill** path (`mla_prefill_ps_asm_fwd` + `mla_reduce_v1`) is gated on
`num_heads % 16 == 0`. Kimi-K3 has 96 heads over `kv_lora_rank=512` → **12 heads/rank at TP8**,
a non-divisor of 16, so its FP8 prefill falls back to the BF16 FMHA decompress path. That fallback
builds a bf16 working set not covered by the FP8 KV-pool accounting and exhausts the activation
arena at long context (K3 OOM'd at ~197k tokens, KV pool <4% used).

This is the **prefill counterpart to #50578** (asm decode pad 12→16). MLA attention is independent
per query head over the shared latent KV, so padding the query heads up to 16 and slicing the
result back is exact.

Rebased onto current `main` (includes merged #50578). Prefill-only diff: **+48 / −11**,
one file (`rocm_aiter_mla.py`).

## Changes

**FP8 prefill enablement (this PR):**
- Relax the `_fp8_prefill_enabled` gate to also allow `0 < num_heads < 16`, and **gate it on FP8 KV**
  (`kv_cache_dtype_str == "fp8"` in the metadata builder; `is_quantized_kv_cache(kv_cache_dtype)`
  in the impl). Without this, a bf16-KV serve would reserve PS workspace unnecessarily.
- In `_mla_fp8_prefill_attn`: replicate-pad Q/K/V to 16 via `AiterMLAHelper.get_mla_padded_q`
  (from #50578), run PS asm prefill + `mla_reduce_v1` at 16 heads, slice output back to `num_heads`.
- Set `num_head_k = max(16, num_heads)` in `_init_fp8_prefill_ps_buffers` and
  `_build_fp8_prefill_ps_metadata` so work/reduce maps match padded tensors — also cuts partial
  tiles (`gcd(16,256)=16` vs `gcd(12,256)=4`), reclaiming **~6 GiB** PS workspace.

Decode padding / Gluon routing is **not** in this PR — that is #50578 (merged).

## Dependencies / related

- **#50578** (merged) — `AiterMLAHelper.get_mla_padded_q` / asm decode for non-divisor heads.
- **ROCm/aiter#4452** (merged `a63ede724`) — 64-bit paged-KV byte offsets for >4 GB addressing.
- Composes with **#48712** (gate PS workspace on fp8 KV) — this PR includes the same gate locally.

## vllm serve command

**Hardware:** 8× MI355X (gfx950), TP8  
**Model:** `moonshotai/Kimi-K3` (mxfp4)

```bash
export VLLM_ROCM_USE_AITER=1
export VLLM_ROCM_USE_AITER_MOE=1
export VLLM_ROCM_USE_AITER_MOE_SITUV2_A8W4=1
export VLLM_USE_BREAKABLE_CUDAGRAPH=0
export AITER_SITUV2_A8W4=1
export SAFETENSORS_FAST_GPU=1
export GPU_ARCHS=gfx950
export HF_HUB_CACHE=/dev/shm/hf-cache HF_HOME=/dev/shm/hf-cache
export VLLM_ENGINE_READY_TIMEOUT_S=3600
export VLLM_EXECUTE_MODEL_TIMEOUT_SECONDS=3600
export VLLM_HTTP_TIMEOUT_KEEP_ALIVE=900
export HIPBLASLT_WORKSPACE_SIZE=32768
export CUBLASLT_WORKSPACE_SIZE=32768

MODEL=/path/to/Kimi-K3   # e.g. /shared_nfs/models/Kimi-K3

vllm serve "$MODEL" --served-model-name moonshotai/Kimi-K3 \
  --host 0.0.0.0 --port 8888 --tensor-parallel-size 8 --async-scheduling \
  --distributed-executor-backend mp \
  --gpu-memory-utilization 0.95 \
  --max-num-seqs 64 --max-model-len 1048576 --max-num-batched-tokens 4096 \
  --trust-remote-code --load-format auto --moe-backend aiter \
  --kv-cache-dtype fp8 --attention-backend ROCM_AITER_MLA --mm-encoder-tp-mode data \
  --compilation-config '{"mode":3,"cudagraph_mode":"FULL_AND_PIECEWISE","custom_ops":["+fused_rms_norm_gated"]}' \
  --enable-prefix-caching --no-disable-hybrid-kv-cache-manager \
  --reasoning-parser kimi_k3 --tool-call-parser kimi_k3 --enable-auto-tool-choice \
  --enable-prompt-tokens-details --disable-uvicorn-access-log

## Test results

### Long-context prefill (functional)

| | Before PR-A | After PR-A |
|---|-------------|------------|
| Fresh prefill @ util 0.95 | OOM ~197k tokens (BF16 FMHA fallback, KV <4%) | **470k (68 s)**, **590k (28.6 s)** OK |
| PS workspace | ~4032 partial tiles (~12 heads) | ~960 partial tiles (~6 GiB saved) |

### Accuracy — GSM8K (lm-eval, full 1319-test split, 5-shot greedy, conc=32, ctx=16384)

Patches applied for this run: **#50578 + PR-A + #50618 only** (no cgmem/MoE/kv_b_proj). Server at
`gpu-memory-utilization 0.88`.

```bash
export HF_HUB_CACHE=/dev/shm/hf-cache HF_HOME=/dev/shm/hf-cache
export OPENAI_API_KEY=EMPTY

python3 -m lm_eval --model local-chat-completions --apply_chat_template \
  --include_path /path/to/InferenceX/utils/evals \
  --tasks /path/to/InferenceX/utils/evals/gsm8k.yaml \
  --output_path /tmp/eval_out --log_samples \
  --model_args "model=moonshotai/Kimi-K3,base_url=http://localhost:8888/v1/chat/completions,api_key=EMPTY,eos_string=</s>,max_retries=5,num_concurrent=32,timeout=1800,tokenized_requests=False,max_length=16384" \
  --gen_kwargs "max_tokens=12288,temperature=0,top_p=1"
```

| Run | n | strict-match EM | flexible-extract EM |
|-----|---|-----------------|---------------------|
| **With PR-A** (MI355X node 193, 2026-08-09) | 1319 | **97.04%** ± 0.47% | **96.97%** ± 0.47% |
| Prior stack (cgmem+MoE+PR-A, node 1319, util 0.95) | 1319 | 96.89% ± 0.48% | 96.89% ± 0.48% |
| Baseline replicate (stock + ASM deps, node 1319) | 1319 | 96.66% ± 0.49% | 96.66% ± 0.49% |

No accuracy regression vs baseline; delta within run noise.

### CODE REVIEW

**Comment by hongxiayang** [comment]
- [ ] vllm serve command
- [ ] eval result

**Comment by hongxiayang** [comment]
@xiaohuguo2023 : Can you resolve the conflict, and update your branch, and re-run and update the PR with full vllm serve command, and also the accuracy  eval  result.

**Comment by xiaohuguo2023** [comment]
> @xiaohuguo2023 : Can you resolve the conflict, and update your branch, and re-run and update the PR with full vllm serve command, and also the accuracy eval result.

@hongxiayang done, thanks!

**Comment by xiaohuguo2023** [comment]
Could anyone help to add the verified label when you have a moment? Pre-commit is gated for new contributors and I’ve run the hooks locally.

cc @hongxiayang @AndreasKaratzas

**Comment by hongxiayang** [comment]
please update your branch as this PR https://github.com/vllm-project/vllm/pull/51011 is merged, and re-run your test. THanks.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 1]
```diff
@@ -1104,6 +1136,9 @@ def _mla_fp8_prefill_attn(
             final_lse,
         )
 
+        if _pad16:
```
Comment (Rohan138) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py]: Is there a way to drop the extra induced copy here btw? Could we maybe have a follow-up to have AITER support the padding+slicing in-kernel?

I'm not sure about initializing `out_3d` here w.r.t CUDAGraph/presistent mla/prefix caching. Maybe we should initialize this as a static memory address in the constructor instead?
cc @dllehr-amd @tjtanaa

Comment (xiaohuguo2023) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py]: Goodquestion

The copy can't drop it from this side. `out` is packed for 12 heads while
the kernels write 16-head rows, so no view of `out` can receive the result,
therrefore we have to use the scratch p+ slice-back. Happy to open the AITER 
follow-up though. In fact, the output copy is the relative small,  ~6 KB/token; 
the q/k/v padding that the same 16-head rule forces costs ~28 KB/token. 
A `num_real_heads` argument would kill all four copies at once, not just this one.

out_3d is now moved to the workspace manager. It now comes from the same
`get_simultaneous` call as `logits`/`attn_lse`/`final_lse`, with its max shape
reserved in `_init_fp8_prefill_ps_buffers` before the workspace is locked. That
gets you the fixed address without a second sizing mechanism — a constructor
buffer would need `max_num_batched_tokens`, which the impl doesn't have but the
builder does. Costs 32 MiB at an 8K token budget, and only when `num_heads < 16`.
Thanks for pointing this out @Rohan138

File: tests/kernels/attention/test_rocm_aiter_mla_fp8_prefill.py [hunk 2]
```diff
@@ -105,6 +105,7 @@ def _build_prefill_metadata(seq_lens: list[int], device: torch.device):
         max_num_reqs=len(seq_lens),
         max_prefill_qlen=max_q,
         max_num_batched_tokens=total_q,
+        attn_out_dtype=torch.bfloat16,
```
Comment (AndreasKaratzas) [inline tests/kernels/attention/test_rocm_aiter_mla_fp8_prefill.py]: why do you have to hard code that here?

Comment (xiaohuguo2023) [reply tests/kernels/attention/test_rocm_aiter_mla_fp8_prefill.py]: Good catch, fixed and  also found a real gap in the non-test path, so just pushed a second commit: _fp8_prefill_enabled now also requires model_config.dtype == torch.bfloat16. Previously an fp16 model with an fp8 KV cache on gfx950 would take this path and get bf16 bits reinterpreted as fp16 — wrong numbers, no error. It now falls back to the bf16 prefill path.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 3]
```diff
@@ -414,9 +414,13 @@ def __init__(
...
-        # FP8 MLA prefill (kn_mla_reduce_v1) only supports 16-aligned heads.
-        self._fp8_prefill_enabled = (
-            _fp8_mla_prefill_supported() and self.num_heads % 16 == 0
+        # The assembly prefill requires FP8 KV, bf16 output, and 16-aligned
+        # heads. It writes bf16 through a raw output pointer, so fp16 must use
+        # the standard prefill path.
+        self._fp8_prefill_enabled = _fp8_mla_prefill_supported() and (
+            kv_cache_dtype_str == "fp8"
+            and vllm_config.model_config.dtype == torch.bfloat16
+            and (self.num_heads % 16 == 0 or 0 < self.num_heads < 16)
```
Comment (shen-shanshan) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py]: When using TP<8 (e.g., TP=4), `num_head` could be 24 and then `_fp8_prefill_enabled` could be false. I'm not sure that should we also enable this kernel for deploy configs other than tp=8?

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 4]
```diff
@@ -1082,7 +1109,19 @@ def _mla_fp8_prefill_attn(
...
         total_q = q.shape[0]
-        nhead = self.num_heads
+        # PS asm prefill + mla_reduce_v1 require 16-aligned heads and the PS
+        # metadata is built for max(16, num_heads). For non-divisor small head
+        # counts (K3 = 12/rank at TP8) replicate-pad q/k/v to 16 — MLA attention
+        # is independent per query head over the shared KV, so the padding heads
+        # cannot affect the real ones (exact, same as the decode path) — then
+        # slice the output back to the real head count.
+        _real_nhead = self.num_heads
+        _pad16 = _real_nhead < 16
```
Comment (shen-shanshan) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py]: Maybe it's better to change all the hardcoded `16` to `AiterMLAHelper._AITER_MIN_MLA_HEADS`, which could make the code more robust.

Comment (xiaohuguo2023) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py]: Done, and it turned out to reach further than the one literal. The same constant was also baked into the PS metadata sizing as max(16, self.num_heads) in _init_fp8_prefill_ps_buffers and build(), and that has to move in lockstep with the forward's pad target — so both now go through one helper, AiterMLAHelper.get_fp8_prefill_num_heads.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py (hunk 3 above)
Comment (xiaohuguo2023) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py]: @shen-shanshan, good catch on the gate being narrower than the kernel needs, But I checked the TP4 case specifically and it doesn't hold up, as K3 can't run with TP4 due to large weights compacity.  

I also surveyed the other MLA models to see if any lands in the rejected band (n > 16 and n % 16 != 0): Kimi-K2 64, LongCat-Flash 64, Kimi-VL 16, DeepSeek-V3/R1 128, All fine at every TP. MiniMax-M2 (48) and GLM-4.6 (96) have head counts that would land there, but both are GQA, not MLA. 96 heads appears to be unique to K3 today

But I think the change is still worth making, for a reason that's stronger than the TP4 one. Head count is a property of the attention config, not of model size, the only thing making 24/rank unreachable is that the single 96-head MLA model also happens to be 1.4 TiB.

**Review by shen-shanshan — APPROVED** [review]
